### PR TITLE
Add precompile contract to deploy root chain staking contract

### PR DESCRIPTION
**Pull request by ninjaahhh**
needed to have a deterministic staking contract address

for more details, check QuarkChain/QCEPs#2

### CODE REVIEW

File: quarkchain/evm/specials.py [hunk 1]
```diff
@@ -254,6 +256,27 @@ def proc_transfer_mnt(ext, msg):
...
+        0,
+        msg.gas - gascost,
+        ROOT_CHAIN_POSW_CONTRACT_BYTECODE,
+        msg.depth + 1,
+        to_full_shard_key=msg.to_full_shard_key,
+        transfer_token_id=msg.transfer_token_id,
+        gas_token_id=msg.gas_token_id,
+    )
+    # CREATE2, deterministic staking contract address
+    return create_contract(ext, new_msg, b"", salt=bytes(32))
```
Comment (qizhou) [inline quarkchain/evm/specials.py]: Why not create a contract with a specific address like precompiled contracts?

File: quarkchain/constants.py [hunk 2]
```diff
@@ -26,3 +26,9 @@
...
+""".strip()
+)
```
Comment (qizhou) [inline quarkchain/constants.py]: I will suggest adding a field ROOT_CHAIN_POSW_CONTRACT_BYTECODE_HASH and use the field to check in shard state

File: quarkchain/evm/specials.py (hunk 1 above)
Comment (ninjaahhh) [reply quarkchain/evm/specials.py]: then the handling logic of the tx targeting at precompiled address will be kinda complicated, something like this:

```
# messages.py

    if msg.code_address in ext.specials:
        res, gas, dat = ext.specials[msg.code_address](ext, msg)
    else:
        res, gas, dat = vm.vm_execute(ext, msg, code)

# specials.py

def proc(ext, msg):
  if ext.has_code(msg.to):
    return vm.vm_execute(ext, msg, code)
  else:
    # make a deployment message
```

which complicates the logic a little bit and I don't see much benefit here, except that the address is slightly easier to remember

File: quarkchain/constants.py (hunk 2 above)
Comment (ninjaahhh) [reply quarkchain/constants.py]: will do

Comment (ninjaahhh) [reply quarkchain/constants.py]: update: I added the bytecode hash to config, but leaves the actual bytecode in the constants

**Pull request merged**